pkg/queue/breaker_test.go

```diff
@@ -79,142 +65,140 @@ func TestBreakerInvalidConstructor(t *testing.T) {
 
 func TestBreakerOverload(t *testing.T) {
 	params := BreakerParams{QueueDepth: 1, MaxConcurrency: 1, InitialCapacity: 1}
-	b := NewBreaker(params)           // Breaker capacity = 2
-	want := []bool{true, true, false} // Only first two requests will be processed
+	b := NewBreaker(params) // Breaker capacity = 2
+	reqs := newRequestor(b)
 
-	locks := b.concurrentRequests(context.Background(), 3)
+	// Bring breaker to capacity.
+	reqs.request()
+	reqs.request()
 
-	unlockAll(locks)
+	// Overshoot by one.
+	reqs.request()
+	reqs.expectFailure(t)
 
-	if diff := cmp.Diff(accepted(locks), want); diff != "" {
-		t.Errorf("Unexpected accepted requests (-want +got): %s", diff)
-	}
+	// The remainer should succeed.
+	reqs.processSuccessfully(t)
+	reqs.processSuccessfully(t)
 }
 
-func TestBreakerOverloadWithEmptySemaphore(t *testing.T) {
-	params := BreakerParams{QueueDepth: 1, MaxConcurrency: 1, InitialCapacity: 0}
-	b := NewBreaker(params)           // Breaker capacity = 2
-	want := []bool{true, true, false} // Only first two requests are processed
+func TestBreakerNoOverload(t *testing.T) {
+	params := BreakerParams{QueueDepth: 1, MaxConcurrency: 1, InitialCapacity: 1}
+	b := NewBreaker(params) // Breaker capacity = 2
+	reqs := newRequestor(b)
 
-	b.sem.release()
-	locks := b.concurrentRequests(context.Background(), 3)
+	// Bring request to capacity.
+	reqs.request()
+	reqs.request()
 
-	unlockAll(locks)
+	// Process one, send a new one in, at capacity again.
+	reqs.processSuccessfully(t)
+	reqs.request()
 
-	if diff := cmp.Diff(accepted(locks), want); diff != "" {
-		t.Errorf("Unexpected accepted requests (-want +got): %s", diff)
-	}
-}
+	// Process one, send a new one in, at capacity again.
+	reqs.processSuccessfully(t)
+	reqs.request()
 
-func TestBreakerNoOverload(t *testing.T) {
-	params := BreakerParams{QueueDepth: 1, MaxConcurrency: 1, InitialCapacity: 1}
-	b := NewBreaker(params)                // Breaker capacity = 2
-	want := []bool{true, true, true, true} // Only two requests will be in flight at a time
-	locks := make([]request, 4)
-	locks[0] = b.concurrentRequest(context.Background())
-	locks[1] = b.concurrentRequest(context.Background())
-	unlock(locks[0])
-	locks[2] = b.concurrentRequest(context.Background())
-	unlock(locks[1])
-	locks[3] = b.concurrentRequest(context.Background())
-	unlockAll(locks[2:])
-
-	if diff := cmp.Diff(accepted(locks), want); diff != "" {
-		t.Errorf("Unexpected accepted requests (-want +got): %s", diff)
-	}
+	// Process the remainder successfully.
+	reqs.processSuccessfully(t)
+	reqs.processSuccessfully(t)
 }
 
 func TestBreakerRecover(t *testing.T) {
```

Comment:
With this makeover I can see much clearer that we do not need this test. E.g. it completely subsumes `TestBreakerOverload`, so it can be let go.
On the other side, I don't see a test where we queue things, update capacity and check things are processed (or additional capacity can be queued) -- we have only capacity checks and some semaphore tests.

Reply:
Added a specific test for queueing, ditched this one.